src/dns.ml

```diff
@@ -1480,6 +1480,27 @@ module Loc = struct
 
 end
 
+(* Null record *)
+module Null = struct
+  type t = bytes
+
+  let to_string null = Hex.show @@ Hex.of_bytes null
+
+  let pp ppf null = Fmt.pf ppf "NULL %s" (to_string null)
+
+  let compare = Bytes.compare
+
+  let decode_exn names buf ~off ~len =
```

Comment:
Why's a function called `decode_exn` returning a `result` type? Shouldn't it raise an exception?

Reply:
It looks like this naming scheme was introduced here: https://github.com/mirage/ocaml-dns/commit/9ad199d8ac01c235c986b43bdee9d4b82ba5b4da

Reply:
> [return a result](https://github.com/mirage/ocaml-dns/blob/1d73cfa96db7452e52b49804d546fad55154aca8/src/dns.ml#L3196-L3253)

I think this could be simplified by replacing the `let*` binding over the Result monad with `let` for those functions that don't return an Error. However perhaps we should leave this to a separate PR and stick with the existing scheme for now.

Reply:
indeed, the naming convention is that decoding functions that may raise are postfixed by `_exn` to illustrate that. And indeed, `Cstruct.BE.get_uint32` can raise if the buffer is smaller than offset + 4.

In this case, both `off` and `len` are passed in (checked) to the function, so how would a realistic example look that leads to an exception? As far as I understand the `Cstruct.to_bytes` documentation, the raise may happen if `off` and/or `len` are provided, and the buffer doesn't meet the size.

Reply:
E.g. `module Ptr` defines a `decode` function since it won't ever raise an exception.

Reply:
> In this case, both `off` and `len` are passed in (checked) to the function, so how would a realistic example look that leads to an exception? As far as I understand the `Cstruct.to_bytes` documentation, the raise may happen if `off` and/or `len` are provided, and the buffer doesn't meet the size.

I suppose if the wrong `off` and `len` where passed? I didn't take into account the callee context when writing this. I'm happy to change it if you prefer though.

Reply:
Ok, then maybe take the `callee context` into account? This function is not exposed, and there's only one call site.

The call site does indeed check `Cstruct.length buf - off >= len` (Rr_map.decode), so as mentioned above the buffer will span at least `off + len` bytes.